### PR TITLE
fix: WebVTT position setting not working

### PR DESCRIPTION
**Pull request by Robloche**
Fixed implementation of specs: https://www.w3.org/TR/webvtt1/#webvtt-cue-position The 'auto' value was missing for position rule, plus text alignment was not used if no alignment was given for position.

Fixes https://github.com/shaka-project/shaka-player/issues/4486

### CODE REVIEW

**Comment by Robloche** [comment]
@joeyparrish Do you need anything else from me in order to proceed with this PR?

**Comment by joeyparrish** [comment]
@Robloche, no, I'm just very far behind on PRs.  Trying to get caught up on bug fix PRs today.

**Comment by joeyparrish** [comment]
I think for this PR, what I really need is to add some layout/screenshot tests that exercise specific subtitle parsers like VTT and TTML.  Then I could:

1. Check that your changes don't break any existing WebVTT layouts
2. Add a regression test that covers this change

Apologies for delaying your work further, but I think tests like those are long overdue.

**Comment by Robloche** [comment]
> I think for this PR, what I really need is to add some layout/screenshot tests that exercise specific subtitle parsers like VTT and TTML. Then I could:
> 
> 1. Check that your changes don't break any existing WebVTT layouts
> 2. Add a regression test that covers this change
> 
> Apologies for delaying your work further, but I think tests like those are long overdue.

Just to be clear, you want me to add tests in `ui_text_displayer_unit.js`, right?

**Comment by joeyparrish** [comment]
> Just to be clear, you want me to add tests in `vtt_text_parser_unit.js` and `ui_text_displayer_unit.js`, right?
> 
> **EDIT:** I wrote some regression tests in `vtt_text_parser_unit.js`. Now, I need help on how to generate official layout screenshots.

Sorry for being unclear!  I was offering to create some basic layout tests and screenshots in our lab that cover existing functionality, then come back to this PR and add a new screenshot.  That way, we not only have a new test for this fix, but we can be sure that it doesn't accidentally change the tests for what was already working.

But the regression unit tests you added are good, too.  Thanks!

**Comment by Robloche** [comment]
My bad! I indeed misunderstood you.
So, my part is done here.

Thanks for the screenshots.

**Comment by Robloche** [comment]
Merry Christmas, @joeyparrish!

Could I have an update on this PR?
If I can do anything to speed up the process, please do tell me.

**Comment by Robloche** [comment]
Any update @joeyparrish?
I don't want to be too pushy but we're waiting for this PR in order to release our next app's version.
Again, if there's anything I can do to help you, feel free to ask.

**Comment by joeyparrish** [comment]
Apologies.  I haven't had time yet.  I promise to prepare new releases as soon as this is resolved.

**Comment by Robloche** [comment]
> Apologies. I haven't had time yet. I promise to prepare new releases as soon as this is resolved.

Thank you so much.
Any idea of timing? Maybe in the following days? Or weeks?

**Comment by joeyparrish** [comment]
I'm working on it now, and I hope to have the first step in review later today.  Here is what to expect:

1. PR to refactor, extracting commonalities in layout test setup and steps
2. PR to add new tests for WebVTT layout

**Comment by joeyparrish** [comment]
I'm testing your PR now against a work-in-progress VTT test suite, and I'm seeing several regressions.  I'll try to get the tests reviewed ASAP so you can see for yourself and iterate on the results.

**Comment by Robloche** [comment]
OK, thanks.

**Comment by joeyparrish** [comment]
@Robloche, we just merged the new end-to-end layout tests.  After running the tests locally, you can review the results on your local web server using the tools at `/test/test/assets/screenshots/review.html`.  I hope this is a useful guide for your work.  Thanks!

**Comment by Robloche** [comment]
I launched the tests then check `review.html` but most of the `-new`and `-diff` images are missing.
image

In fact, I only see the screenshots for Windows/Chrome and Windows/Edge.
Am I missing something, here?

My second issue is about the differences.
The red area says there are differences but I don't think I changed any code related to this (on top of these differences being very slight):
image image

Could you please enlight me about these tests?

**Comment by joeyparrish** [comment]
> I launched the tests then checked `review.html` but most of the `-new`and `-diff` images are missing.
> In fact, I only see the screenshots for Windows/Chrome and Windows/Edge. Am I missing something, here?

No, that's correct.  If you only run the tests on Windows/Chrome and Windows/Edge, only those will have -new and -diff images, because only those were run.  This is normal.  Use the filters at the top to show only one or both of the platforms you have.

> My second issue is about the differences. The red area says there are differences but I don't think I changed any code related to this (on top of these differences being very slight)

Those are often sub-pixel rendering differences around edges, or slight color variations for your graphics card and/or monitor.  You can and should ignore those, and those tests should still be passing.

Please use the filters in review.html to select only the tests that fail or that you are working on.

**Comment by Robloche** [comment]
Thanks, that's clear.
I'm working on a totally unrelated thing at the moment but I'll try to find time to address this as soon as possible.

**Comment by joeyparrish** [comment]
What I saw with your PR was that the positioning in the new webvtt tests was off.  You should merge main in and try those new tests.

**Comment by Robloche** [comment]
I merged main and now I see the issues...

**Comment by Robloche** [comment]
@joeyparrish 
I spent quite some time, digging into the specs, Shaka's code and my fix proposal.
I found some errors in my implementation but, I mostly found errors in the tests.

For instance, take this one:
image

This is the screenshot that should be matched to successfully pass the test.
But clearly, it's wrong because, I you can see, the cue box is centered around the 50% mark.
I should be centered around the 90% mark, like `position:90%`.
I read multiple times the specs at https://www.w3.org/TR/webvtt1/#webvtt-cue-position-alignment and I really think it should be displayed like this:
image

Of course, if I'm right, some of the tests should be updated to match the specs.

I made a quick demo that you could play with at: https://codesandbox.io/s/pedantic-chatterjee-cr1enf
It embeds a version of Shaka Player that includes my fixes.

Please have a look and let me know what you think.

**Comment by Robloche** [comment]
There are also issues with the following tests: `align-end`, `align-end-long`, `align-start` and `align-start-long`.
Like the specs says, if text alignment is `left` (resp. `right`), the `auto` position translates to `0%` (resp. `100%`):
image

But if text alignment if `start` or `end`, the `auto` position translates to `50%`:
image

More over, since no position alignement is given, it falls back to `auto`, and the specs tells us that this translates to `line-left` (resp. `line-right`) when text alignment is `start` (resp. `end`):
image

And this is obviously not what the current tests show, since both cue boxes are centered at 50%:
image image

What I would expect here, is:
image image

**Comment by joeyparrish** [comment]
Oh, there are definitely issues prior to your PR.  As I created that test suite based on the upstream VTT web platform tests, I left comments on the tests where we clearly did the wrong thing.  We needed a baseline to start from, to keep us aware of changes, and to make sure that future PRs are only making things better and never worse.

**Comment by Robloche** [comment]
OK. I understand.
So what are the next steps?

Tell me if I can be of any help.

**Comment by joeyparrish** [comment]
> So what are the next steps?

Double-check your changes against the test suite.  If you look in `test/text/web_vtt_layout_integration.js`, you'll see `FIXME` comments detailing what we know is wrong with our rendering of those cases.  Whatever you change, check that it's both correct and an improvement, and update/delete those comments as appropriate.  When you're satisfied with the work you've done, we'll review the results.

I just built a GitHub Actions workflow (PRs #4991 and #4993) to update a PR with screenshots from all platforms in our test lab.  It's still a bit rough, but mostly working, and I can trigger that for you at any time to update all the changed screenshots.

**Comment by joeyparrish** [comment]
Also, thank you again for working on this, and for being patient with us while we build these tools to make sure we are moving in the right direction!  Changes in subtitle rendering have been painful for us in the past without these tools, and regressions were extremely common.  You are one of the first to go down this path with us.

**Comment by Robloche** [comment]
I'm not sure what I'm supposed to do...

The following UI tests are OK with my fixes (hence the FIXME comments should be removed):
https://github.com/shaka-project/shaka-player/blob/ac963a9d1ac78ebd22eee6df49c48e520f9c9969/test/text/web_vtt_layout_integration.js#L94
https://github.com/shaka-project/shaka-player/blob/ac963a9d1ac78ebd22eee6df49c48e520f9c9969/test/text/web_vtt_layout_integration.js#L106
https://github.com/shaka-project/shaka-player/blob/ac963a9d1ac78ebd22eee6df49c48e520f9c9969/test/text/web_vtt_layout_integration.js#L118
https://github.com/shaka-project/shaka-player/blob/ac963a9d1ac78ebd22eee6df49c48e520f9c9969/test/text/web_vtt_layout_integration.js#L130
https://github.com/shaka-project/shaka-player/blob/ac963a9d1ac78ebd22eee6df49c48e520f9c9969/test/text/web_vtt_layout_integration.js#L155

For this one, the UI test is OK but the Native one is still incorrect:
https://github.com/shaka-project/shaka-player/blob/ac963a9d1ac78ebd22eee6df49c48e520f9c9969/test/text/web_vtt_layout_integration.js#L180

**Comment by Robloche** [comment]
@joeyparrish Any update on the tests?

**Comment by joeyparrish** [comment]
> I'm not sure what I'm supposed to do...
> 
> The following UI tests are OK with my fixes (hence the FIXME comments should be removed):

Then please remove/update those comments and upload your latest changes.  Let us know when you're ready for review.  I'll run the workflow to update your PR with the latest screenshots on all platforms, and I'll review the diffs.

**Comment by Robloche** [comment]
> > I'm not sure what I'm supposed to do...
> > The following UI tests are OK with my fixes (hence the FIXME comments should be removed):
> 
> Then please remove/update those comments and upload your latest changes. Let us know when you're ready for review. I'll run the workflow to update your PR with the latest screenshots on all platforms, and I'll review the diffs.

I can remove them, but I cannot update them since I don't know how to proceed to fix the invalid test images.

**Comment by Robloche** [comment]
@joeyparrish Could you please clarify what I'm supposed to do?

**Comment by joeyparrish** [comment]
> I can remove them, but I cannot update them since I don't know how to proceed to fix the invalid test images.

I will run an automated workflow on GitHub that will update your PR with new images on all platforms.  Please let us know when you are ready for that.

**Comment by Robloche** [comment]
Well, I am ready.

**Comment by joeyparrish** [comment]
Okay, thanks.  I'm using the lab for another PR at the moment, but I'll run yours next.

**Comment by joeyparrish** [comment]
Your job is now in progress: https://github.com/shaka-project/shaka-player/actions/runs/4356666987

File: lib/text/ui_text_displayer.js [hunk 1]
```diff
@@ -644,30 +716,73 @@ shaka.text.UITextDisplayer = class {
...
-      if (cue.writingMode == Cue.writingMode.HORIZONTAL_TOP_TO_BOTTOM) {
-        style.paddingLeft = cue.position;
+    // TODO: factorize code since width/height, left/top and right/bottom
+    //       have same role, depending on the writing mode
+    if (cue.writingMode === Cue.writingMode.HORIZONTAL_TOP_TO_BOTTOM) {
+      // Horizontal writing
+      if (computedPositionAlign === Cue.positionAlign.LEFT) {
+        // Left side of cue box is positioned at 'computedPosition'
+        style.paddingLeft = computedPosition + '%';
+        style.width = (100 - computedPosition) + '%';
```
Comment (joeyparrish) [inline lib/text/ui_text_displayer.js:739]: Is padding and width really the right way to align the cue box?  Padding is internal to the element, and width includes padding.

So if the element has a solid background, that background covers the padding area, too.

Also, since width includes padding, if computedPosition is 50%, you end up with width=50%, padding=50%.  If width == padding, there's no room for content, right?

File: lib/text/ui_text_displayer.js [hunk 2]
```diff
@@ -644,30 +716,73 @@ shaka.text.UITextDisplayer = class {
...
+          // Center of cue box is in the right part of the viewport
+          const width = (100 - computedPosition) * 2;
+          style.width = width + '%';
+          style.paddingLeft = (100 - width) + '%';
+        } else if (computedPosition < 50) {
+          // Center of cue box is in the left part of the viewport
+          const width = computedPosition * 2;
+          style.width = width + '%';
+          style.paddingRight = (100 - width) + '%';
+        }
```
Comment (joeyparrish) [inline lib/text/ui_text_displayer.js:757]: Shouldn't computedPosition==50 be handled in one of the above two cases?  Is it right to leave out styles in this case?

File: lib/player.js [hunk 3]
```diff
@@ -7026,7 +7026,7 @@ shaka.Player.TYPICAL_BUFFERING_THRESHOLD_ = 0.5;
  * @define {string} A version number taken from git at compile time.
  * @export
  */
-shaka.Player.version = 'v4.3.0-uncompiled';
+shaka.Player.version = 'v4.4.0-uncompiled';
```
Comment (avelad) [inline lib/player.js:7029]: Why?

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "shaka-player",
   "description": "DASH/EME video player library",
-  "version": "4.3.0",
+  "version": "4.4.0",
```
Comment (avelad) [inline package.json:4]: Why?

File: package-lock.json [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
   "name": "shaka-player",
-  "version": "4.3.0",
+  "version": "4.4.0",
   "lockfileVersion": 2,
   "requires": true,
   "packages": {
     "": {
       "name": "shaka-player",
-      "version": "4.3.0",
+      "version": "4.4.0",
```
Comment (avelad) [inline package-lock.json:9]: Why? Revert this change

**Review by avelad — CHANGES_REQUESTED** [review]
Please remove .idea files

File: lib/text/ui_text_displayer.js [hunk 6]
```diff
@@ -644,30 +716,73 @@ shaka.text.UITextDisplayer = class {
       }
     }
 
-    style.lineHeight = cue.lineHeight;
+    const computedPositionAlign = this.computeCuePositionAlignment_(cue);
```
Comment (joeyparrish) [inline lib/text/ui_text_displayer.js:719]: GitHub won't let me leave comments on binary files, so here are my takes on the screenshots further down the PR.  Please correct me if I'm misinterpreting WebVTT for any of these.

 - webvtt-ui-align-center-position-10.png: I don't believe it should be squished into two lines.  I suspect that is a consequence of the width setting below.
 - webvtt-ui-align-center-position-90.png: Same.
 - webvtt-ui-align-end-long.png: This is very clearly a regression, IMO.  The text flow should not change from what it was.  The only issue with the previous version was that the background is meant to hug the text, not fill the cue box.
 - webvtt-ui-align-end.png: This should be aligned on the right (according to Chrome's native implementation), but it actually moved left of center in this PR.
 - webvtt-ui-bold-long.png and many other non-alignment tests: These all moved to the right, with the left edge centered in the screen for some reason.  I don't think this is right.

**Review by joeyparrish — CHANGES_REQUESTED** [review]
I think there's a fundamental problem with the width+padding model of alignment you've implemented, and it seems to affect a large number of test cases negatively.  None of these effects are platform-specific, and so you didn't need to wait for me to test them in the lab to see the effects.

File: lib/player.js (hunk 3 above)
Comment (Robloche) [reply lib/player.js:7029]: I don't know how this happened. I'll revert.

File: package.json (hunk 4 above)
Comment (Robloche) [reply package.json:4]: I don't know how this happened. I'll revert.

File: package-lock.json (hunk 5 above)
Comment (Robloche) [reply package-lock.json:9]: I don't know how this happened. I'll revert.

File: .idea/.gitignore [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Default ignored files
```
[the remark this answers is not in the document]
Comment (Robloche) [reply .idea/.gitignore:1]: I don't want to include them. I removed them.

File: lib/player.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Robloche) [reply lib/player.js:7029]: Done.

File: package-lock.json [hunk 8]
```diff
@@ -1,12 +1,12 @@
 {
   "name": "shaka-player",
-  "version": "4.3.0",
+  "version": "4.4.0",
```
[the remark this answers is not in the document]
Comment (Robloche) [reply package-lock.json:3]: Done.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,99 @@
 # Changelog
 
+## [4.4.0](https://github.com/shaka-project/shaka-player/compare/v4.3.0...v4.4.0) (2023-02-06)
```
[the remark this answers is not in the document]
Comment (Robloche) [reply CHANGELOG.md:3]: It must have been pulled when I updated my branch with upstream?

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (Robloche) [reply package.json:4]: Done.

**Comment by Robloche** [comment]
> I think there's a fundamental problem with the width+padding model of alignment you've implemented, and it seems to affect a large number of test cases negatively. None of these effects are platform-specific, and so you didn't need to wait for me to test them in the lab to see the effects.

I think my latest changes have not been pushed to this PR because when I run the tests on my computer, the only faulty tests I see are outdated tests. I'll check my recent commits...

File: lib/text/ui_text_displayer.js (hunk 1 above)
Comment (Robloche) [reply lib/text/ui_text_displayer.js:739]: This padding is not applied to the element containing the text, but to the container of this element.
So, for example, on the image below, you can see a padding of 50% *and* a width of 50%:
image

File: lib/text/ui_text_displayer.js (hunk 2 above)
Comment (Robloche) [reply lib/text/ui_text_displayer.js:757]: In this particular case, the center of the box is centered at 50% of the viewport. So, there's really nothing to change to the style.
I might have missunderstood the specs, of course, but what I see in my tests seems valid.

File: lib/text/ui_text_displayer.js (hunk 6 above)
Comment (Robloche) [reply lib/text/ui_text_displayer.js:719]: > * webvtt-ui-align-center-position-10.png: I don't believe it should be squished into two lines.  I suspect that is a consequence of the width setting below.
> * webvtt-ui-align-center-position-90.png: Same.

Allow me to disagree. If the line is not to wrap on two lines, then what are the possibilities?
1. Display on one line and truncate the text.
2. Display the entire text on one line and don't respect the position settings.

I don't like both of these solutions and I've found 2 resources that are aligned with me (I'm not cherry-picking, I didn't find any other page):

- https://ronallo.com/demos/webvtt-cue-settings/ (drag the position cue to 10 and see what happens).
- https://www.speechpad.com/captions/webvtt
`Cue payloads are the subtitle text that is to be displayed on the screen. They come immediately after the cue timing line, and can be one or more lines in length. Line separators will be preserved. That is, even if the combined text of two lines would fit on one line, the subtitle will still be presented on two lines as indicated in the file. However, if a line will not fit the width of the cue box (aka caption box), it will be wrapped, not truncated.`

Plus if you look at example 1 at https://www.w3.org/TR/webvtt1/#introduction-multiple-lines, wrapping text on multiple lines seems to be the philosophy:
image


> * webvtt-ui-align-end-long.png: This is very clearly a regression, IMO.  The text flow should not change from what it was.  The only issue with the previous version was that the background is meant to hug the text, not fill the cue box.

Again, I disagree. Without any position setting, it defaults to 50%. So the right side of the cue box should be positioned at 50% of the viewport. But I admit my implementation suffers from the same "hug" issue.


> * webvtt-ui-align-end.png: This should be aligned on the right (according to Chrome's native implementation), but it actually moved left of center in this PR.

`align:end` means that the text is right-aligned in the cue box.
Since there's no position setting, it defaults to 50.
There's no position alignment setting as well, so applying the algorithm [here](https://www.w3.org/TR/webvtt1/#webvtt-cue-position-alignment) give us the computed position alignment `line-right`.
These 3 settings gives us a cue box that is right-aligned at 50% of the viewport, with a text right-align inside it.


> * webvtt-ui-bold-long.png and many other non-alignment tests: These all moved to the right, with the left edge centered in the screen for some reason.  I don't think this is right.
Same reason as above: when no position or position alignment is given, their value are computed as described in the specs.


I think at least one of us does not understand the specs, and that could totally be me.
But reading them again and again does not make me understand them in another way.

All in all, I want to thank you for your time, and truly hope we'll finally align (right or left, whatever).

**Comment by Robloche** [comment]
I fully understand that you have another agenda (one that actually feeds you), no worries.
Is there any possibility that someone else have a look at this issue and at the specs?
Maybe we're both wrong, after all.

> One possible path forward is for me to spend some time trying to fix these test cases in a different way, informed by this PR. I'll let you know what I can do when I have time to look at this again. In the meantime, we welcome other perspectives and alternate approaches from the rest of the community.

OK for that.

File: lib/text/vtt_text_parser.js [hunk 10]
```diff
@@ -398,7 +399,7 @@ shaka.text.VttTextParser = class {
       cue.endTime = end;
       cue.payload = '';
     } else {
-      cue = new shaka.text.Cue(start, end, '');
+      cue = new Cue(start, end, '');
```
Comment (avelad) [inline lib/text/vtt_text_parser.js:402]: This change is not necessary

File: lib/text/vtt_text_parser.js [hunk 11]
```diff
@@ -888,8 +889,10 @@ shaka.text.VttTextParser = class {
       cue.positionAlign = Cue.positionAlign.LEFT;
     } else if (align == 'line-right' || align == 'end') {
       cue.positionAlign = Cue.positionAlign.RIGHT;
-    } else {
+    } else if (align == 'center') {
```
[the remark this answers is not in the document]
Comment (Robloche) [reply lib/text/vtt_text_parser.js:892]: I added `middle`.

File: lib/text/vtt_text_parser.js (hunk 10 above)
Comment (Robloche) [reply lib/text/vtt_text_parser.js:402]: OK. I reverted it.

**Comment by avelad** [comment]
@Robloche can you rebase and resolve the conflicts? Thanks!

**Comment by avelad** [comment]
@Robloche can you fix the build errors? Thanks!

File: lib/text/vtt_text_parser.js [hunk 12]
```diff
@@ -792,9 +792,9 @@ shaka.text.VttTextParser = class {
     } else if ((results = /^size:([\d.]+)%$/.exec(word))) {
       cue.size = Number(results[1]);
     } else if ((results =
-        // eslint-disable-next-line max-len
-        /^position:([\d.]+)%(?:,(line-left|line-right|middle|center|start|end|auto))?$/
-            .exec(word))) {
+      // eslint-disable-next-line max-len
```
Comment (avelad) [inline lib/text/vtt_text_parser.js:795]: this change is not necessary, please revert it

**Comment by avelad** [comment]
@Robloche can you rebase and resolve the conflicts? Thanks!
